WIP Feature/wds 21 add pytest black mypy

jamesoutterside opened the pull request:
- Added pytest, black and mypy & resolved related issues. 
- Updated dev docker setup + test. 
- Updated model bakery 
- Starts some of the settings changes but that's anther task  so will likely need full review/redo, only fixes important key for now.

rolo commented on .gitignore:
```diff
@@ -6,4 +6,19 @@ __pycache__/
 *~
 
+# cookie cutter
+{{cookiecutter.project_name}}/requirements-dev.txt
```
I'm not a fan of this multiple requirements pattern. It's a slightly faster build time in exchange for a slighy difference in environments and an extra file in the repo. I appreciate it's now the defacto in the front end world. Happy to be outvoted on this one. Interested in what everyone else thinks.

jamesoutterside replied:
I don't have a strong argument for or against, imo it's tidier and as you say removes some bloat on deployment/potentially removes some chance of developer error like leaving debug tools enabled by mistake etc but those aren't huge reasons, so happy to go with the consensus. 

I actually really like the way DRF does it https://github.com/encode/django-rest-framework/tree/master/requirements but guessing that is even worse from your perspective!

OmegaDroid replied:
I've been playing with this myself, I have requirements.txt with

```
-r requirements-server.txt
-r requirements-test.txt
```

and both of those get built by pip-tools. In my main container I install `requirements-server.txt` and I have a test container that extends the server that also installs `requirements-test.txt`, for dev your workflow doesn't change and you just install `requirements.txt` / `pip-sync`.

The only issue with this approach is if there is a dependency mismatch, for instance if package `A` requires and version of `B` in the server reqs but the `C` requires `B<X` there will be a miss match when trying to install `requirements.txt` but this then just requires adding the restriction to both `.in` files.

Personally I think it's worth it for both image size and build time.

radiac replied:
I lean gently towards splitting dev out. On PHE NSCR we removed quite a bit of bloat - 52 deployed packages against 125 dev packages, which seems sensible.

Almost more importantly, it forces a decision about where a package goes, and so should help clarify how important things are when it comes to managing upgrades etc.

Main negative it's also easy to add a package to the wrong requirement in dev, and wonder why it's failing on CI.

I like DRF's idea of grouping requirements, but think it'd be more manageable to keep them in the one file, then group them and explain their purpose with comments.

Agree with Rolo about ending up with more files though - when you end up with lots of requirements it can get confusing. Think it's worth keeping the number of requirements files to a minimum, eg `requirements.in` and `requirements-dev.in` (which itself includes `requirements.in`, to enforce dev consistency with deployed).

OmegaDroid replied:
Id be fine with that other than changing it so dev is in `requirements.in` and there is a `requirements-deployed.in` that way when devving there's nothing special and installing specific reqs file is stored in script.

jamesoutterside replied:
As discussed, single in for now added.

rolo commented on .gitignore:
```diff
@@ -6,4 +6,19 @@ __pycache__/
 {{cookiecutter.project_name}}/requirements.txt
+{{cookiecutter.project_name}}/static/images/
+
+# Editor files
```
I feel like someone's local editor files belong in their local global git ignore file. Not that I can recall where that is right now, or if we do that for other projects. Don't feel super strongly on this one.

jamesoutterside replied:
Yep that's fair. It's also already in the .gitignore in the CC, so will review what's in there, there is probably a 'standard' python one that could be used, will have a look.

radiac replied:
Also this isn't a complete list. Our other .gitignores are quite cluttered with these sort of things.

That said, perhaps better exclude them here than accidentally get one merge? Also don't feel strongly.

jamesoutterside replied:
How about going with https://github.com/github/gitignore/blob/master/Python.gitignore (for both the template and the cut project) then it's fairly standard. 

Note that that doesn't include environments, so that's likely a better idea to have locally as you said.

jamesoutterside replied:
Used http://gitignore.io/api/django for now, probably a good place to then refine from, there is some stuff we likely don't need in there.

rolo commented on README.rst:
```diff
@@ -5,7 +5,7 @@ Wildfish Django Starter
     :target: https://travis-ci.org/wildfish/wildfish-django-starter
     
-A Django 2.0 friendly project cookiecutter template to kick start development for new projects.  Includes apps and settings we use in the majority of projects, along with an integrated version of our other cookiecutter-django-crud template which will also generate a model, CRUD views and tests.
+A Django 2.2 friendly project cookiecutter template to kick start development for new projects.  Includes apps and settings we use in the majority of projects, along with an integrated version of our other cookiecutter-django-crud template which will also generate a model, CRUD views and tests.
```
I think let's just take out the version so we don't forget to update it. Change to:

"A Django project cookiecutter"

jamesoutterside replied:
Yep, added.

rolo commented on {{cookiecutter.project_name}}/requirements-dev.in:
```diff
@@ -0,0 +1,31 @@
+sphinx_rtd_theme
+
+# Deployment
+fabric
```
We should probably remove fabric in favour of invoke.

radiac replied:
I believe invoke isn't a straight replacement for fabric, but is used alongside fabric 2? The `upgrade` branch has fabric 2, probably best to address this there.

jamesoutterside replied:
I was wondering, with this and the testing deployments bit in upgrade, with us splitting out k8s etc. Should build and deployment also be split to another project repo or is there still value having it in here?

radiac replied:
Lets make it js- and deployment-agnostic for now - strip out anything about building js, fabric, deployment; we can add them back in when we need them.

jamesoutterside replied:
Removed invoke and fabric.

rolo commented on {{cookiecutter.project_name}}/{{cookiecutter.project_name}}/settings.py:
```diff
@@ -1,124 +1,167 @@
+    return value.split(",")
 
 
 class RedisCache(object):
```
Lets trash this class and combine with the Dev settings.

rolo replied:
Can come after merge.

jamesoutterside replied:
Yeah, planning on going over the whole settings file in a separate PR/issue. Only did what was needed this one.

radiac replied:
Merge with Deployed class

radiac commented on setup.cfg:
```diff
@@ -1,3 +1,2 @@
 [tool:pytest]
-python_paths = .
-norecursedirs = .tox .git */migrations/* */static/* docs venv */{{cookiecutter.project_name}}/*
+testpaths = tests
```
Do we need more here? See:
* https://github.com/PublicHealthEngland/nsc-recommendations/blob/develop/setup.cfg#L3-L5
* https://github.com/radiac/django-fastview/blob/develop/setup.cfg#L52-L53

jamesoutterside replied:
Ah yeah, will need adopts, was just to get it working. That's the top level one not the CC one, which has that all in, so won't need the Django bits.

jamesoutterside replied:
Added, I've left coverage off, the test will cover the cookie_cutter so this would only be covering the tests themselves.

OmegaDroid commented on {{cookiecutter.project_name}}/dev-docker-compose.yml.default:
```diff
@@ -0,0 +1,42 @@
+      - "5432:5432"
+    volumes:
+      - "../db:/db"
+      - "../backup:/backup"
```
would be good to have the db and backup paths configurable by env var, I have a separate drive for storing all my data for example. Something like `"${PG_DB_DATA:-../db}:/db"` would be cool

OmegaDroid replied:
also, wouldn't those default paths clash with other projects assuming you checkout all projects to a single root dir

radiac replied:
Yes, this assumes it's checked out to `project/repo`; my standard project structure is
```
project
  repo
  venv
  files
  notes.rst
```
so made sense for me to add `backup` for somewhere to put the sample db. Again, env var would solve this?

jamesoutterside replied:
Also removed build if that's project specific.

OmegaDroid commented on {{cookiecutter.project_name}}/dev-docker-compose.yml.default:
```diff
@@ -0,0 +1,42 @@
+
+services:
+  postgres:
+    image: postgres
```
Not sure if we should be pinning here? may be best to leave that to a per project basis or have it use an env var that defaults to latest

rolo replied:
I'd go with no pinning. It's only a starting point.

jamesoutterside replied:
Using an env to use latest by default, which can be overridden.

OmegaDroid commented on {{cookiecutter.project_name}}/dev-docker-compose.yml.default:
```diff
@@ -0,0 +1,42 @@
+      PGDATA: /db
+
+  redis:
+    image: redis
```
same as pg for pinning to a specific version

jamesoutterside replied:
Same fix as above.

OmegaDroid commented on {{cookiecutter.project_name}}/dev-docker-compose.yml.default:
```diff
@@ -0,0 +1,42 @@
+    environment:
+      DJANGO_CONFIGURATION: Dev
+      DATABASE_HOST: postgres
+      DJANGO_SECRET_KEY: {{cookiecutter.secret_key}}
```
Id rather not have the secret key written into the repo at all, this could be any string and we can generate a secret key for each deployed env and store that as a secret

radiac replied:
Agreed - secret key is now optional in settings for dev

OmegaDroid commented on {{cookiecutter.project_name}}/{{cookiecutter.project_name}}/settings.py:
```diff
@@ -200,13 +253,11 @@ class Prod(Deployed):
+            "NAME": get_env("POSTGRES_USER", ""),
+            "USER": get_env("POSTGRES_USER", ""),
+            "PASSWORD": get_env("POSTGRES_PASSWORD", "password"),
+            "HOST": get_env("POSTGRES_SERVICE_HOST", "localhost"),
```
This should probably be in the base and just override the settings in Dev

jamesoutterside replied:
Will takle this one in the settings task.

jamesoutterside commented:
Merged in settings branch, so this can be checked again all together. Couple of notes: 

- Added settings/splice of the PHE and Crypdit. 
- Black 120  is not here yet, needs a pyproject.toml (I think) and that's an open discussion elsewhere I believe and it leads to some not-so-great formatting of dicts, magic comma solution (at least for logging) didn't work. 
- Added a separate task to discuss settings security values. 
- Logging is pinned back/simpler for now. 
- Redis is a mixin before deployed so it can be used for the travis tests (test coverage for cache is needed). 
- Removed the docker test - changing it to actually used postgres caused a wait issue, attempted to resolve with a few docker suggestions but no luck, can get it to wait successfully for postgres to start(and then actually start) without a hacked sleep in the test_docker.sh file.

radiac commented on tox.ini:
```diff
@@ -0,0 +1,26 @@
+    TOXENV={envname}
+    DJANGO_REDIS_HOST: localhost
+    py{36,37,38}-django2.2: TOX_DJANGO_VERSION=django>=2.2, <3.0
+    py{36,37,38}-django3.0: TOX_DJANGO_VERSION=django>=3.0, <3.1
```
Does it make sense for us to test multiple versions of django here?

jamesoutterside replied:
It was put in before the 2/3 discussion, could probably be stripped back to 3/latest now if that's the plan?

radiac commented on {{cookiecutter.project_name}}/{{cookiecutter.project_name}}/settings.py:
```diff
@@ -1,124 +1,208 @@
+    """
+    if value is None:
+        return []
+    return value.split(",")
```
If we want to use these three fns, would it be worth splitting them out into their own project to simplify tests and maintainability?
